### PR TITLE
8310241: OffsetDateTime compareTo redundant computation

### PR DESCRIPTION
**Pull request by RogerRiggs**
Remove a redundant comparison in java.time `OffsetDateTime.compareTo()`. 
If the `compareInstant` utility method returns 0 (equal), it compares the `LocalDateTime`.
However, `compareInstant` has already done that comparison; if it found equal, the `compareTo` method unnecessarily does it again.
The code is refactored in `compareTo` to do the comparison of `LocalDateTime` exactly once, if it is needed.

This case is NOT covered by an existing test in test/jdk/java/time/tck/java/time/TCKOffsetDateTime.java

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310241](https://bugs.openjdk.org/browse/JDK-8310241): OffsetDateTime compareTo redundant computation (**Bug** - P4)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14618/head:pull/14618` \
`$ git checkout pull/14618`

Update a local copy of the PR: \
`$ git checkout pull/14618` \
`$ git pull https://git.openjdk.org/jdk.git pull/14618/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14618`

View PR using the GUI difftool: \
`$ git pr show -t 14618`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14618.diff">https://git.openjdk.org/jdk/pull/14618.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14618#issuecomment-1603183221)

### CODE REVIEW

File: src/java.base/share/classes/java/time/OffsetDateTime.java [hunk 1]
```diff
@@ -167,17 +167,26 @@ public static Comparator<OffsetDateTime> timeLineOrder() {
      * Compares this {@code OffsetDateTime} to another date-time.
      * The comparison is based on the instant.
      *
+     * When two values represent the same instant, the local date-time is compared
+     * to distinguish them. This step is needed to make the ordering
+     * consistent with {@code equals()}.
+     *
```
Comment (naotoj) [inline src/java.base/share/classes/java/time/OffsetDateTime.java:173]: This method is also called from the public `timeLineOrder()` method. Could this change cause a different order?

Comment (RogerRiggs) [reply src/java.base/share/classes/java/time/OffsetDateTime.java:173]: Your are right; I missed that subtlety. With that in mind I don't think the performance impact is sufficient to duplicate more code.

**Comment by RogerRiggs** [comment]
Closing because this 'fix' a bigger issue than it fixes.

**Comment by RogerRiggs** [comment]
Reopening this to propose a different implementation and to 
provide the test of the specific condition that is missing from java/time/jck/java/time/TCKOffsetDateTime.java

File: test/jdk/java/time/tck/java/time/TCKOffsetDateTime.java [hunk 2]
```diff
@@ -1381,6 +1381,7 @@ public void test_compareTo_hourDifference() {
         assertEquals(a.compareTo(a) == 0, true);
         assertEquals(b.compareTo(b) == 0, true);
         assertEquals(a.toInstant().compareTo(b.toInstant()) == 0, true);
+        assertEquals(OffsetDateTime.timeLineOrder().compare(a,b) == 0, true);
```
Comment (naotoj) [inline test/jdk/java/time/tck/java/time/TCKOffsetDateTime.java:1384]: Could be in a separate test case, as this method tests `compareTo` method?

Comment (RogerRiggs) [reply test/jdk/java/time/tck/java/time/TCKOffsetDateTime.java:1384]: This is one of a sequence of tests of compareTo.  The other (previous) tests include verifying the assertion about the result of the `timeLineOrder` comparator with the same data.

**Review by naotoj — APPROVED** [review]
LGTM

**Pull request closed without merging**